ScarletEngine/include/Core/Memory/Memory.h

```diff
@@ -0,0 +1,70 @@
+#pragma once
+
+#include "MemoryTracker.h"
+#include <iostream>
+namespace ScarletEngine
+{
+	template <class T>
+	struct GlobalAllocator
+	{
+		using size_type = size_t;
+		using difference_type = std::ptrdiff_t;
+		using pointer = T*;
+		using const_pointer = const T*;
+		using reference = T&;
+		using const_reference = const T&;
+		using value_type = T;
+
+		GlobalAllocator() = default;
+		template <class U> constexpr GlobalAllocator(const GlobalAllocator <U>&) {}
+
+		[[nodiscard]] T* allocate(size_t Number) 
+		{
+			T* Ptr = static_cast<T*>(std::malloc(Number * sizeof(T)));
+			check(Ptr != nullptr);
+			MemoryTracker::MarkAlloc(Ptr, Number * sizeof(T));
+			return Ptr;
+		}
+
+		void deallocate(T* Ptr, size_t Number)
+		{
+			check(Ptr != nullptr);
+			MemoryTracker::RemoveAlloc(Ptr, Number * sizeof(T));
+			std::free(Ptr);
+		}
+
+		template <typename... Args>
+		[[nodiscard]] static T* New(Args&&... args)
```

Comment:
I read up on no discard but when is the best practice to use it?

Reply:
No discard should be used when you want to statically assert (**compiler** will throw and error) that the return value of your function is not being discarded. For example here we would never want to call new and then just simply discard the return value (this would cause leaked memory since we would have no way to delete it after).
Discard in this context means that you never _use_ the return value of the function

```cpp
int* a = GlobalAllocator<int>(); // Good
GlobalAllocator<int>(); // bad, leaked memory, the compiler error because you discarded the return value
```

There are niche uses for `MakeShared` which don't need to be `[[nodiscard]]` but its safer to just enforce for all cases.

Reply:
Best practice is that it should probably be almost everywhere but because its ugly af (you could `#define NODISCARD [[nodiscard[]]` but meh) I only put it where its actually a logical error to discard the return value.